planner/cascades/transformation_rules.go

```diff
@@ -589,3 +592,40 @@ func (r *PushSelDownJoin) OnTransform(old *memo.ExprIter) (newExprs []*memo.Grou
 	newJoinExpr.SetChildren(leftGroup, rightGroup)
 	return []*memo.GroupExpr{newJoinExpr}, true, false, nil
 }
+
+// PushTopNDownProjection pushes TopN to Projection.
+type PushTopNDownProjection struct {
+	baseRule
+}
+
+// NewRulePushTopNDownProjection creates a new Transformation PushTopNDownProjection.
+func NewRulePushTopNDownProjection() Transformation {
+	rule := &PushTopNDownProjection{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandTopN,
+		memo.EngineTiDBOnly,
+		memo.NewPattern(memo.OperandTopN, memo.EngineTiDBOnly),
+	)
+	return rule
+}
+
+// OnTransform implements Transformation interface.
+// This rule tries to pushes the TopN through Projection.
+func (r *PushTopNDownProjection) OnTransform(old *memo.ExprIter) (newExprs []*memo.GroupExpr, eraseOld bool, eraseAll bool, err error) {
+	topN := old.GetExpr().ExprNode.(*plannercore.LogicalTopN)
+	proj := old.Children[0].GetExpr().ExprNode.(*plannercore.LogicalProjection)
+	childGroup := old.Children[0].GetExpr().Children[0]
+	for _, by := range topN.ByItems {
+		by.Expr = expression.ColumnSubstitute(by.Expr, proj.Schema(), proj.Exprs)
+	}
+	// remove meaningless constant sort items.
+	for i := len(topN.ByItems) - 1; i >= 0; i-- {
+		switch topN.ByItems[i].Expr.(type) {
+		case *expression.Constant, *expression.CorrelatedColumn:
+			topN.ByItems = append(topN.ByItems[:i], topN.ByItems[i+1:]...)
+		}
+	}
+	topNExpr := memo.NewGroupExpr(topN)
+	topNExpr.SetChildren(childGroup)
+	return []*memo.GroupExpr{topNExpr}, true, false, nil
```

Comment:
The pattern we have matched is `TopN->Projection->X`. And we want to transform it to `Projection->TopN->X`. So here, we should:
1. Create a new `GroupExpr`: projExpr which contains the unmodified `Projection`.
2. Create a new `GroupExpr`: `topNExpr` which contains the new `TopN`.
3. Set `topNExpr`'s children to `childGroup`.
4. Create a `Group`: `topNGroup` which contains `topNExpr`.
5. Set `projExpr`'s children to `topNGroup`.
6. return `[]*memo.GroupExpr{projExpr}, true, false, nil`.

Reply:
Ohhh. I got, does it should like this?
```
	projExpr := memo.NewGroupExpr(proj)
	topNExpr := memo.NewGroupExpr(topN)
	topNExpr.SetChildren(childGroup)
	topNGroup := memo.NewGroupWithSchema(topNExpr, topN.Schema())
	projExpr.SetChildren(topNGroup)
	return []*memo.GroupExpr{projExpr}, true, false, nil
```

Reply:
1. `topNGroup` should use the schema of `childGroup.Schema`.
2. The `topN` cannot be the origin `topN`. You should create a new `LogicalTopN`. We'd better assume all of the `GroupExpr` is immutable.(BTW, I think the `ByItems` should be cloned from the old TopN)